spatialmath/capsule.go

```diff
@@ -0,0 +1,346 @@
+	if dist > CollisionBuffer {
+		return capsuleVsMeshDistance(c, other.toMesh())
+	}
+	return dist
+}
+
+// IMPORTANT: meshes are not considered solid. A mesh is not guaranteed to represent an enclosed area. This will measure ONLY the distance
+// to the closest triangle in the mesh.
+func capsuleVsMeshDistance(c *capsule, other *mesh) float64 {
+	lowDist := math.Inf(1)
+	for _, t := range other.triangles {
+		// Measure distance to each mesh triangle
+		dist := capsuleVsTriangleDistance(c, t)
+		if dist < lowDist {
+			lowDist = dist
+		}
+	}
+	return lowDist
+}
+
+func capsuleVsTriangleDistance(c *capsule, other *triangle) float64 {
+	capPt, triPt := closestPointsSegmentTriangle(c.segA, c.segB, other)
+	return capPt.Sub(triPt).Norm() - c.radius
+}
+
+func capsuleVsBoxCollision(c *capsule, other *box) bool {
+	// TODO(pl): Large amounts of capsule collision code were adopted from `brax`
+	// https://github.com/google/brax/blob/7eaa16b4bf446b117b538dbe9c9401f97cf4afa2/brax/physics/colliders.py
+	// https://github.com/google/brax/blob/7eaa16b4bf446b117b538dbe9c9401f97cf4afa2/brax/physics/geometry.py
+	// Brax converts boxes to meshes composed of 12 triangles and does collision detection on those.
+	// We support doing so but it appears that separating axis test is superior so that is what is used here.
+	// The two methods should be fully characterized.
+	return capsuleBoxSeparatingAxis(c, other)
+}
+
+// capsuleInCapsule returns a bool describing if the inner capsule is fully encompassed by the outer capsule.
+func capsuleInCapsule(inner, outer *capsule) bool {
+	return capsuleVsPointDistance(outer, inner.segA) < -inner.radius &&
+		capsuleVsPointDistance(outer, inner.segB) < -inner.radius
+}
+
+// capsuleInBox returns a bool describing if the given capsule is fully encompassed by the given box.
+func capsuleInBox(c *capsule, b *box) bool {
+	return pointVsBoxDistance(c.segA, b) <= -c.radius && pointVsBoxDistance(c.segB, b) <= -c.radius
+}
+
+// capsuleInSphere returns a bool describing if the given capsule is fully encompassed by the given sphere.
+func capsuleInSphere(c *capsule, s *sphere) bool {
+	return c.segA.Sub(s.pose.Point()).Norm()+c.radius <= s.radius && c.segB.Sub(s.pose.Point()).Norm()+c.radius <= s.radius
+}
+
+// capsuleBoxSeparatingAxis returns immediately as soon as any result is found indicating that the two objects are not in collision.
+func capsuleBoxSeparatingAxis(a *capsule, b *box) bool {
+	capCenter := a.segA.Add(a.segB).Mul(0.5)
+	centerDist := b.pose.Point().Sub(capCenter)
+
+	// check if there is a distance between bounding spheres to potentially exit early
+	if centerDist.Norm()-((a.length/2)+b.boundingSphereR) > CollisionBuffer {
+		return false
+	}
+	rmA := a.pose.Orientation().RotationMatrix()
+	rmB := b.pose.Orientation().RotationMatrix()
+
+	segLen := (a.length - 2*a.radius) / 2
+	capVec := rmA.Row(2).Mul(segLen)
```

Comment:
this is hard to follow.  Looks like youre getting the plane bisecting the capsule here?  could you add a comment?

Reply:
Comment was added